### PR TITLE
First pass at sniff to confirm inline docs for filters are present

### LINKED ISSUE #424: Check that hooks follow documentation standards
Forked from https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/241#issuecomment-133562703.

### PR DESCRIPTION
**Pull request by kraftbj**
Fixes #424

First pass at sniff to confirm inline docs for hooks are present.

I'm adding this into Jetpack via https://github.com/Automattic/jetpack/pull/15693 since I wanted to check for some Jetpack-specific things. Per suggestion, wanted to offer the "base" to core WPCS for everyone's benefit.

In the initial commit, it finds a matching the various hook function (L27) using WPCS' `AbstractFunctionRestrictionsSniff` and checks for a preceding comment. The preceding comment need to be on the previous line (noting this is an open question from the Jetpack PR https://github.com/Automattic/jetpack/pull/15693#pullrequestreview-406406725) and confirm it is using the docblock format.

For the initial commit, it is verifying it has a `@since` tag within the docblock and that the value starts with a X.Y.Z version number. I'm not sure about WP itself, but Jetpack requires three levels to both confirm the exact version (e.g. 8.5 vs 8.5.0) and ensure that all hooks in the same major release (X.Y) will have the same value to avoid our docs parser tagging it as two separate versions.

If Core WPCS doesn't need that, I can remove that to just verify a `@since` tag is present and move the version check itself to Jetpack as I did with our `@module` tag in https://github.com/Automattic/jetpack/pull/15693/files#diff-eb7bc4afcdf23e6f99ae1f793afd119bR57

### CODE REVIEW

**Comment by jrfnl** [comment]
Hi @kraftbj Welcome to WPCS ;-)

Thanks for this PR. I haven't looked at the code yet, but would like to ask you whether you have run the sniff over WP Core to see what it throws up.

Any false positives from WP Core should be handled, or when in doubt whether something is a false positive, let's discuss it.

> For the initial commit, it is verifying it has a `@since` tag within the docblock and that the value starts with a X.Y.Z version number. I'm not sure about WP itself, but Jetpack requires three levels to both confirm the exact version (e.g. 8.5 vs 8.5.0) and ensure that all hooks in the same major release (X.Y) will have the same value to avoid our docs parser tagging it as two separate versions.

Core requires three levels too as per the [Documentation guidelines](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#since-section-changelogs), so checking for three levels is fine.

FYI: I expect WPCS 3.0.0 to contain a trait for version number checking (written for another sniff), so that can probably be re-used.

With that in mind and the last 2.x release expected soon, after which the pulls for WPCS 3.0.0 will start, I'd like to earmark this PR for WPCS 3.0.0.

**Comment by kraftbj** [comment]
Sounds good. New to PHPCS overall, so appreciate the guidance on naming scheme. Took a guess that `WP` seem to fit best based on the existing directories.

I'll get some unit tests worked up early next week since it sounds like this is a sniff that would be considered for inclusion.

**Comment by jrfnl** [comment]
> New to PHPCS overall, so appreciate the guidance ...

Please feel free to ask for as much guidance as you need from me and the others in the team.

In my case, my remarks will often presume you know what you're doing with PHPCS (as why would you send in a sniff otherwise?), please don't let that intimidate you and ask for as much clarification as you need.

> it sounds like this is a sniff that would be considered for inclusion.

Definitely. There are long standing open issues about this check missing from WPCS, though I also have to caution you that it may not be the easiest issue to tackle. Can be a great way to learn more about writing sniffs though ;-)

**Comment by kraftbj** [comment]
Sounds good. I appreciate the tip on running it against WP to find false positives. I'll make a couple of tweaks when sending up the unit tests.

**Comment by GaryJones** [comment]
> For the initial commit, it is verifying it has a `@since` tag within the docblock and that the value starts with a X.Y.Z version number.

Maybe `@since Unknown` should also be permitted? I believe that has/had some instances in core function/methods DocBlocks as well at one point ([example in SimplePie package](https://github.com/WordPress/WordPress/search?q=%22%40since+Unknown%22&unscoped_q=%22%40since+Unknown%22)).

**Comment by jrfnl** [comment]
> Maybe @since Unknown should also be permitted?

I don't think we should allow that. Both SVN as well as Git allow to trace things through history, so I can't think of a valid reason to allow `@since Unknown`.

**Comment by kraftbj** [comment]
I added unit tests and documentation, but I'm not sure how to unit test when the ruleset would be passing a property as an array.

With the X.Y.Z. format, there are two instances currently in Core which does not follow this which I am not keen at automatically changing:
* `0.71` -- the initial release did not follow the X.Y.Z format.
* `MU (3.0.0)` -- to denote items that came in when MU was merged into Core. We could modify these to be `3.0.0 (MU)` as an alternative.

To allow these, I set a rule property of an array that could be defined in the ruleset XML for any "exceptional" versions. I just don't know how to pass that to the unit test via `phpcs:set`. 

Any pointers?

**Comment by kraftbj** [comment]
> Maybe `@since Unknown` should also be permitted?

For Core, I would disagree, but for implementing projects, I did add a way to set an `exceptional` version. If an implementing project wanted to allow that, they could in their own `phpcs.xml.dist` file as the PR stands now.

**Comment by jrfnl** [comment]
> I'm not sure how to unit test when the ruleset would be passing a property as an array.

See: https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress/Tests/NamingConventions/PrefixAllGlobalsUnitTest.1.inc#L6 for an example of setting an array property.

Please also remember to reset the property after the test using it and/or at the end of the test file (see bottom of the above linked test file for an example).

**Comment by kraftbj** [comment]
@jrfnl I really appreciate the pointers and direction. I think this is at a state where you can ~rip it to shreds~ review it.

**Comment by jrfnl** [comment]
@kraftbj Wowie! I'll try and have a look some time this week ;-)

File: WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+ * @package WPCS\WordPressCodingStandards
+ */
+class HooksInlineDocsSniff extends AbstractFunctionRestrictionsSniff {
+
+	/**
+	 * Array of WordPress hook execution functions.
+	 *
+	 * @var array WordPress hook execution function name => filter or action.
+	 */
+	protected $hook_functions = array(
```
Comment (GaryJones) [inline WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php:27]: We have a lot of arrays of functions names in the `Sniff` class that all of the Abstract (and therefore concrete) classes extend from - might be worth renaming the property to a bit more explicitly (`$hook_executing_functions`?) and then moving them to there so then other sniffs that need this collection can use them.

File: WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+		$previous_comment = $this->return_previous_comment( $stack_ptr );
+
+		if ( false !== $previous_comment ) {
+			/*
+			 * Check to determine if there is a comment immediately preceding the function call.
+			 */
+			if ( ( $this->tokens[ $previous_comment ]['line'] + 1 ) !== $this->tokens[ $stack_ptr ]['line'] ) {
+				$this->phpcsFile->addError(
+					'The inline documentation for a hook must be on the line immediately before the function call.',
```
Comment (GaryJones) [inline WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php:100]: As noted previously, this may not always be possible. Consider this example from Juliette:

```
/**
 * Hook docblock
 */
if ( $a === true
    && apply_filter('some_filter', $value) > 10
) {
    // Do something.
}
```

Likewise, this should also be valid IMO, since it matches the PSR-5 definition of associating a DocBlock with a Structural Element with an optional number of clear lines in between:

```
/**
 * Hook docblock
 */

$value = apply_filters( 'some_filter', $value );
```

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+// Has a since tag with an incomplete value.
+/**
+ * This is a description.
+ *
+ * @since 8.5
+ */
+apply_filters( 'hook', 'function' );
+
+// Has a since tag with an wrong value.
```
Comment (GaryJones) [inline WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc:53]:
```suggestion
// Has a since tag with a wrong value.
```

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+apply_filters( 'hook', 'function' );
+
+// "Already documented" example of an action.
+/** This action is documented in file.php */
+do_action( 'hook ' );
+
+// "Already documented" example of a filter.
+/** This filter is documented in file.php */
+apply_filters(' hook', 'function' );
+
```
Comment (GaryJones) [inline WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc:107]: These are good, but some real-world edge cases will be needed!

Consider:

```php
/**
 * This is my function DocBlock.
 *
 * @since 0000.00000.0000000000
 */
function my_function_has_a_hook() {
    do_action( 'foo' );
}
```

Likewise:

```php
/**
 * This is my function DocBlock.
 *
 * @since 0000.00000.0000000000
 */
function my_function_has_a_hook() {
    echo 'foo';
}
do_action( 'foo' );
```

If these both already get reported as missing the hook DocBlock, then great! Add them in as unit tests, and then try:

```php
/**
 * This is my function DocBlock.
 *
 * @since 0000.00000.0000000000
 */
function my_function_has_a_hook() { do_action( 'foo' ); }
```

😈

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc [hunk 5]
```diff
@@ -0,0 +1,108 @@
+<?php
+// phpcs:set WordPress.Commenting.HooksInlineDocs allowed_extra_versions[] MU
```
Comment (GaryJones) [inline WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc:2]: This doesn't have to be specified at the top - you could move it to just before it's needed, so that you can check that the behaviour works correctly without any allowed_extra_versions having been set.

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+<documentation title="Hooks Inline Docblock Standard">
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:1]: b50e8e3d

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    <standard>
+        <![CDATA[
+    All WordPress hook execution functions should be preceded by a docblock explaining the hook.
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:4]: b50e8e3d

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    ]]>
+    </standard>
+    <code_comparison>
+        <code title="Valid: using a docblock to describe a hook.">
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:8]: b50e8e3d

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            <![CDATA[
+/**
+ * Description of hook.
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:11]: b50e8e3d

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @since 1.0.0
+ */
+do_action( 'hook' );
+        ]]>
+        </code>
+        <code title="Invalid: using a non-docblock comment">
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:18]: b50e8e3d

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * <em>@since 1.0.0</em>
+ */
+do_action( 'hook' );
+        ]]>
+        </code>
+        <code title="Invalid: including a @since tag with a different undeclared version.">
+            <![CDATA[
+/**
+ * Description of hook.
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:41]: b50e8e3d

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ */
+do_action( 'hook' );
+        ]]>
+        </code>
+    </code_comparison>
+    <code_comparison>
+        <code title="Valid: including a @since tag with a valid X.Y.Z version.">
+            <![CDATA[
+/**
+ * Description of hook.
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:31]: 56960143

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @since 1.0.0
+ */
+do_action( 'hook' );
+        ]]>
+        </code>
+        <code title="Invalid: using a non-docblock comment">
+            <![CDATA[
+/*
+ * This is not a docblock.
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:21]: 56960143

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc (hunk 3 above)
Comment (kraftbj) [reply WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc:53]: b50e8e3d

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc (hunk 5 above)
Comment (kraftbj) [reply WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc:2]: b2855e67

File: WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php (hunk 1 above)
Comment (kraftbj) [reply WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php:27]: Nice pointer. In 11e3a8fc, I removed my array. There is already a `$hookInvokeFunctions` array in the Sniff class.

The difference is the value of each key in the existing one is `true` whereas mine had `filter|action`. Originally, I was going to ensure that the "previously documented" check (`This filter is documented`) used `filter` or `action` correctly, but opted against being that nit-picky.

Since I wasn't used the key, let's just use what's already there.

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc (hunk 4 above)
Comment (kraftbj) [reply WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc:107]: The first two did catch "correctly", but probably wouldn't after I resolve the preceding line issue mentioned in https://github.com/WordPress/WordPress-Coding-Standards/pull/1892#discussion_r424599130

But, the third example of a single-line function fails the unit test since it doesn't see that as an error.

I'm going to push up what I have (and mark this as resolved since the tests are added) and think on it a bit.

I'm thinking for the third example, we can search for a comment previous to the hook but stop at a function declaration at least. May be others.

Thanks!

Comment (kraftbj) [reply WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc]: ad220fdb

File: WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php (hunk 1 above)
Comment (GaryJones) [reply WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php]: If the existing values were all `true`, then chances are the values weren't being used elsewhere. If there is benefit in having them be either 'filter' or 'action' so that _this_ sniff can check a little more finely (would you propose to throw a different violation if they didn't match?), then I'm all for that.

cc @jrfnl for consideration. Unresolving this comment so it doesn't get missed.

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc (hunk 4 above)
Comment (GaryJones) [reply WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc]: Either here or PHPCSUtils, I bet @jrfnl has examples / helper functions to find something within a specific scope i.e. find a DocComment within the current scope.

File: WordPress/Docs/Commenting/HooksInlineDocsStandard.xml [hunk 14]
```diff
@@ -15,7 +15,7 @@
 do_action( 'hook' );
         ]]>
         </code>
-        <code title="Invalid: using a non-docblock comment">
+        <code title="Invalid: not using a DocBlock comment">
```
Comment (GaryJones) [inline WordPress/Docs/Commenting/HooksInlineDocsStandard.xml:20]: A "DocBlock" is made up of two parts:

- the `/**  */` structure, known as a _DocComment_.
- the contents with Summary, optional description, tags, etc. known as the _PHPDoc_.

As such, the "DocBlock comment" here, which is referencing whether one asterisk or two is used, is really checking if it's a multi-line comment (`/*`) or a _DocComment_ (`/**`), hence my earlier suggestion of `Invalid: not using a DocComment."

(Note that the period is still missing at the end from your version.)

File: WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php (hunk 1 above)
Comment (jrfnl) [reply WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php]: I'm pretty sure that the values in the array aren't currently used, but a search of the codebase can verify this for sure.

I'm all for it to have `filter|action` as the values. The error messages could then potentially also use that term if it makes the error message clearer.

File: WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc (hunk 4 above)
Comment (jrfnl) [reply WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc]: I do have docblock finder helper functions lined up to go into PHPCSUtils at some point, but as hook docblocks are a bit of a different beast compared to most structural elements I don't think that will help here.

I do have a draft `HooksTrait` lined up for WPCS 3.0.0 with logic for finding hook docblocks (for another sniff).

When I pull that, I will combine the logic for finding hook docblocks from this sniff with the logic in the trait to get the best of both brains in the trait (and switch this sniff over to using the trait).

**Comment by kraftbj** [comment]
This is wonderful. Thank you so much for the review. Today is a standard holiday in the U.S. so a short week for me. I'll try to make a good dent on this during the week, but may get bumped to next. If so, my delay doesn't mean I'm discouraged! 🙂

**Comment by jrfnl** [comment]
While it would be nice to have a sniff for this, the PR as-is is not mergable with way too many open review comments and no action taken on them in over two years, so I'm going to close this PR.
